src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java

```diff
@@ -0,0 +1,246 @@
+import com.couchbase.client.java.error.DocumentAlreadyExistsException;
+import com.couchbase.client.java.error.subdoc.MultiMutationException;
+import com.couchbase.client.java.subdoc.DocumentFragment;
+
+/**
+ * A CouchbaseArraySet is a {@link Set} backed by a {@link Bucket Couchbase} document (more
+ * specifically a {@link JsonArrayDocument JSON array}).
+ *
+ * Note that as such, a CouchbaseArraySet is restricted to the types that a {@link JsonArray JSON array}
+ * can contain. JSON objects and sub-arrays can be represented as {@link JsonObject} and {@link JsonArray}
+ * respectively. Similarly, null entries are not supported.
+ *
+ * @param <T> the type of values in the set.
+ */
+public class CouchbaseArraySet<T> extends AbstractSet<T> {
+
+    public static final int MAX_OPTIMISTIC_LOCKING_ATTEMPTS = CouchbaseArrayList.MAX_OPTIMISTIC_LOCKING_ATTEMPTS;
+
+    private final String id;
+    private final Bucket bucket;
+
+    /**
+     * Create a new {@link CouchbaseArraySet}, backed by the document identified by <code>id</code>
+     * in the given Couchbase <code>bucket</code>. Note that if the document already exists,
+     * its content will be used as initial content for this collection. Otherwise it is created empty.
+     *
+     * @param id the id of the Couchbase document to back the set.
+     * @param bucket the {@link Bucket} through which to interact with the document.
+     */
+    public CouchbaseArraySet(String id, Bucket bucket) {
+        this.id = id;
+        this.bucket = bucket;
+
+        try {
+            this.bucket.insert(JsonArrayDocument.create(id));
+        } catch (DocumentAlreadyExistsException e) {
+            //use a pre-existing document
+        }
+    }
+
+    /**
+     * Create a new {@link CouchbaseArraySet}, backed by the document identified by <code>id</code>
+     * in the given Couchbase <code>bucket</code>. Note that if the document already exists, its content is
+     * reset to the values copied from the given <code>data</code> Map.
+     *
+     * A null or empty data map will re-initialize any pre-existing document to an empty content.
+     *
+     * @param id the id of the Couchbase document to back the set.
+     * @param bucket the {@link Bucket} through which to interact with the document.
+     * @param initialData Set of the elements to initially store in the CouchbaseArraySet.
+     */
+    public CouchbaseArraySet(String id, Bucket bucket, Set<? extends T> initialData) {
+        this.id = id;
+        this.bucket = bucket;
+
+        JsonArray data = JsonArray.create();
+        if (initialData != null && !initialData.isEmpty()) {
+            for (Object o : data) {
+                data.add(o);
+            }
+        }
+        bucket.upsert(JsonArrayDocument.create(id, data));
+    }
+
+    @Override
+    public int size() {
+        //TODO use subdoc GET_COUNT when available
+        JsonArrayDocument current = bucket.get(id, JsonArrayDocument.class);
```

Comment:
I wonder if stuff like this and empty() all of those that are similar could be refactored into a helper class? so once GET_COUNT and friends come along we only fix it in one place?